src/exchange.ts

```diff
@@ -0,0 +1,131 @@
+      // Essentially, the return value is a bunch of useless values like
+      // {
+      //   id: id,
+      //   symbol: "UNKNOWN",
+      //   decimals: -1,
+      //   address: 0x00....0,
+      // }
+      tokenInfo = {
+        id: id,
+        symbol: fallbackSymbolName,
+        decimals: fallbackDecimals,
+        address: tokenAddress,
+      };
+    }
+    tokenObjects.set(id, tokenInfo);
+    // Found Token ${tokenInfo.symbol} at ID ${tokenInfo.id} with ${tokenInfo.decimals} decimals`
+  }
+  return tokenObjects;
+}
+
+/**
+ * A handy function providing fee token liquidity by placing orders selling the fee token
+ * at the specified provision price for each specified token.
+ * @param exchange - BatchExchange Smart Contract
+ * @param tokenIds - An array of token indices as represented on the exchange
+ * @param provisionPrice - Price at which liquidity is to be provided
+ * @param sellAmountOwl - Amount of feeToken to be sold
+ * @param artifacts - A context-like object providing a gateway to Truffle contract ABIs.
+ * @returns Void Promise
+ */
+export async function placeFeeTokenLiquidityOrders(
+  exchange: BatchExchangeInstance,
+  tokenIds: number[],
+  provisionPrice: BN,
+  sellAmountOwl: BN,
+  artifacts: Truffle.Artifacts,
+): Promise<void> {
+  const minBuyAmounts = [];
+  const validTokenIds = [];
+  const feeToken = await fetchTokenInfoFromExchange(exchange, [0], artifacts);
+  // This is expected to always be OWL which has 18 digits.
+  const feeDigits = feeToken.get(0)?.decimals || 18;
+  const tokenInfo = await fetchTokenInfoFromExchange(
+    exchange,
+    tokenIds,
+    artifacts,
+  );
+  for (const tokenId of tokenIds) {
+    const numDigits = tokenInfo.get(tokenId)?.decimals;
+    if (numDigits && numDigits != -1) {
+      validTokenIds.push(tokenId);
+      if (numDigits < feeDigits) {
+        minBuyAmounts.push(
+          sellAmountOwl
+            .mul(provisionPrice)
+            .div(new BN(10).pow(new BN(feeDigits - numDigits))),
+        );
+      } else {
+        minBuyAmounts.push(
+          sellAmountOwl
+            .mul(provisionPrice)
+            .mul(new BN(10).pow(new BN(numDigits - feeDigits))),
+        );
+      }
+    }
+  }
+  const numOrders = validTokenIds.length;
+  const batchId = (await exchange.getCurrentBatchId()).toNumber();
+  if (numOrders == 0) {
+    // No orders added since all tokens sufficiently funded or not found.
+    return;
+  }
+  await exchange.placeValidFromOrders(
+    validTokenIds, // buyTokens
+    Array(numOrders).fill(0), // sellTokens
+    Array(numOrders).fill(batchId + 2), // validFroms (all to begin 2 batches from now)
+    Array(numOrders).fill(MAXU32), // validTos
+    minBuyAmounts, // buyAmounts
+    Array(numOrders).fill(sellAmountOwl), // sellAmount
+  );
+  // "Placed liquidity sell orders for the following tokens",
+  // await Promise.all(
+  //   validTokenIds.map(async (i) => await tokenInfo.get(i)?.address),
+  // );
```

Comment:
Ist this going to be deleted or do you want to put back the log statement?

Reply:
I want to put a log...